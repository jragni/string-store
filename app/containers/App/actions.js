/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  POST_STRING,
  POST_STRING_SUCCESS,
  POST_STRING_ERROR,
  LOAD_STRINGS,
  LOAD_STRINGS_SUCCESS,
  LOAD_STRINGS_ERROR,
  LOAD_REPOS,
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS_ERROR,
} from './constants';

// TODO: DELETE STARTING HERE
/**
 * Load the repositories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_REPOS
 */
export function loadRepos() {
  return {
    type: LOAD_REPOS,
  };
}

/**
 * Dispatched when the repositories are loaded by the request saga
 *
 * @param  {array} repos The repository data
 * @param  {string} username The current username
 *
 * @return {object}      An action object with a type of LOAD_REPOS_SUCCESS passing the repos
 */
export function reposLoaded(repos, username) {
  return {
    type: LOAD_REPOS_SUCCESS,
    repos,
    username,
  };
}

/**
 * Dispatched when loading the repositories fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_REPOS_ERROR passing the error
 */
export function repoLoadingError(error) {
  return {
    type: LOAD_REPOS_ERROR,
    error,
  };
}
// END DELETE

// TODO: Docstrings for 3 functions below.
/** loadStrings
 *
 */
export function loadStrings() {
  return {
    type: LOAD_STRINGS,
  };
}

/** stringsLoaded
 * TODO: add docstrings
 */
export function stringsLoaded(strings) {
  return {
    type: LOAD_STRINGS_SUCCESS,
    strings,
  };
}

/** stringsLoadingError
 * TODO: add docstrings
 */
export function stringLoadingError(error) {
  return {
    type: LOAD_STRINGS_ERROR,
    error,
  };
}

/**
 * TODO: add docstring
 * @returns
 */
export function postString() {
  return {
    type: POST_STRING,
  };
}

/** stringPosted
 * TODO: add docstrings
 */
export function stringPosted(string) {
  return {
    type: POST_STRING_SUCCESS,
    string,
  };
}

/** stringsLoadingError
 * TODO: add docstrings
 */
export function stringPostingError(error) {
  return {
    type: POST_STRING_ERROR,
    error,
  };
}
