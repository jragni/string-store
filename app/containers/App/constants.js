/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

// TODO: REMOVE UPON COMPLETION
export const LOAD_REPOS = 'boilerplate/App/LOAD_REPOS';
export const LOAD_REPOS_SUCCESS = 'boilerplate/App/LOAD_REPOS_SUCCESS';
export const LOAD_REPOS_ERROR = 'boilerplate/App/LOAD_REPOS_ERROR';
// end TODO

export const LOAD_STRINGS = 'stringstore/App/LOAD_STRINGS';
export const LOAD_STRINGS_SUCCESS = 'stringstore/App/LOAD_STRINGS_SUCCESS';
export const LOAD_STRINGS_ERROR = 'stringstore/App/LOAD_STRINGS_ERROR';

export const POST_STRING = 'stringstore/App/POST_STRING';
export const POST_STRING_SUCCESS = 'stringstore/App/POST_STRING_SUCCESS';
export const POST_STRING_ERROR = 'stringstore/App/POST_STRING_ERROR';
