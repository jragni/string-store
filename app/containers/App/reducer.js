/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { toast } from 'react-toastify';
import {
  LOAD_STRINGS,
  LOAD_STRINGS_SUCCESS,
  LOAD_STRINGS_ERROR,
  POST_STRING,
  POST_STRING_SUCCESS,
  POST_STRING_ERROR,
  LOAD_REPOS_SUCCESS, // TODO: REMOVE
  LOAD_REPOS, // TODO: REMOVE
  LOAD_REPOS_ERROR, // TODO: REMOVE
} from './constants';

// The initial state of the App
// TODO: add initial states to initialState
export const initialState = {
  string: '', // NOTE: added for dev.
  strings: [], // NOTE: added for dev.
  loading: false,
  error: false,
  currentUser: false, // TODO: remove when done
  userData: {
    repositories: false, // TODO: remove when done
  },
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_STRINGS:
        draft.loading = true;
        draft.error = false;
        break;

      case LOAD_STRINGS_SUCCESS:
        draft.loading = false;
        draft.strings = action.strings; // TODO: verify if this works with getting strings
        break;

      case LOAD_STRINGS_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;

      case POST_STRING:
        draft.loading = true;
        draft.error = false;
        break;

      case POST_STRING_SUCCESS:
        draft.loading = false;
        draft.string = action.string; // TODO: verify if this works with getting strings
        toast.success('String added!');
        break;

      case POST_STRING_ERROR:
        draft.error = action.error;
        draft.loading = false;
        toast.error(action.error);
        break;

      case LOAD_REPOS:
        draft.loading = true;
        draft.error = false;
        draft.userData.repositories = false;
        break;

      case LOAD_REPOS_SUCCESS:
        draft.userData.repositories = action.repos;
        draft.loading = false;
        draft.currentUser = action.username;
        break;

      case LOAD_REPOS_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;
    }
  });

export default appReducer;
