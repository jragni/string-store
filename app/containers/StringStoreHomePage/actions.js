/**
 * StringStoreHomePage Actions
 */

const { CHANGE_STRING, SUBMIT_STRING } = require('./constants');

/** changeString
 * Action that changes the input field of the form.
 *
 * @param {string} string: the new text of the input field.
 *
 * @return {object} An action object with type of CHANGE_STRING.
 */
export function changeString(string) {
  return {
    type: CHANGE_STRING,
    string,
  };
}

export function submitString(string) {
  return {
    type: SUBMIT_STRING,
    string,
  };
}
