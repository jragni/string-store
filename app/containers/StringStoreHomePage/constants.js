/** StringStoreHomePage constants
 * Constants used to avoid weird typos between the reducer and Actions.
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'stringstore/StringStoreHomePage/ACTION'
 */

export const CHANGE_STRING = 'stringstore/StringStoreHomePage/CHANGE_STRING';
export const SUBMIT_STRING = 'stringstore/StringStoreHomePage/SUBMIT_STRING';
