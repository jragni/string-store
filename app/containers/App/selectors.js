/**
 * The global state selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const selectRouter = state => state.router;

// NOTE: ADDED FOR DEV
const makeSelectStrings = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.strings,
  );

// END DEV
// TODO: delete unnecessary selectors upon completion
const makeSelectCurrentUser = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.currentUser,
  );

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectRepos = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.userData.repositories,
  );

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

export {
  selectGlobal,
  makeSelectLoading,
  makeSelectError,
  makeSelectStrings,
  makeSelectCurrentUser,
  makeSelectRepos,
  makeSelectLocation,
};
