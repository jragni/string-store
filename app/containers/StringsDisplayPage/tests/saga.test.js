/**
 * Tests for StringDisplayPage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { stringsLoaded, stringsLoadingError } from 'containers/App/actions';
import { LOAD_STRINGS } from '../../App/constants';

import serverData, { getStrings } from '../saga';

// const strings = [{ id: 1, message: 'TEST' }];

/* eslint-disable redux-saga/yield-effects */
describe('getStrings Saga', () => {
  let getStringsGenerator;

  beforeEach(() => {
    getStringsGenerator = getStrings();

    const selectDescriptor = getStringsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should dispatch the stringsLoaded action if it requests the data successfully', () => {
    const response = [
      {
        id: 1,
        message: 'TEST',
      },
    ];

    const putDescriptor = getStringsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(stringsLoaded(response)));
  });

  it('should dispatch the stringsLoadingError action if the response is errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getStringsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(stringsLoadingError(response)));
  });
});

describe('serverData Saga', () => {
  const getServerData = serverData();

  it('should start task to watch for LOAD_STRINGS action', () => {
    const takeLatestDescriptor = getServerData.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_STRINGS, getStrings));
  });
});
