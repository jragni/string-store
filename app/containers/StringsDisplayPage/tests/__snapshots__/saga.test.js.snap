// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStrings Saga should dispatch the stringsLoaded actiohn if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/api/strings",
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
        "methods": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getStrings Saga should dispatch the stringsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/api/strings",
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
        "methods": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getStrings Saga should dispatch the stringsLoadingError action if the response is errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/api/strings",
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
        "methods": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
